qi-lib/flow/core/deforest.rkt

```diff
@@ -0,0 +1,405 @@
+#lang racket/base
+
```

Comment:
Deforestation is by far the most difficult to understand part of this PR for me. I think this file deserves a big 100-line introductory comment explaining the compilation strategy, maybe with an example, and the structure of the code that follows. If there is already an up-to-date description on the wiki a link there would be okay. But as is I'm not sure this file is maintainable other than by Dominik without a lot of effort. I have not tried to review it in detail.

Reply:
Dominik did document this in the wiki here: https://github.com/drym-org/qi/wiki/The-Compiler#optimizations . I haven't reviewed it in detail so I confess I'm not sure re: maintainability, but on first glance it seems comprehensive. Some notes though:

- That wiki doesn't include examples and I agree they would be nice to have
- I second that it would be useful to write a comment at the top of the module explaining the structure of the code that follows, and maybe even signpost each section, like ";; Core stream implementations" ";; Syntax classes to match optimizable input patterns"
- I don't immediately recall exactly what `make-fine-curry` and `make-producer-curry` do -- the comments explain that they make curry 🍛  but they don't say for what purpose or where it's used. 

This module is pretty dense and could benefit from more signposting/putting the code in context. Wdyt @dzoep ?

For now I'll go ahead and add a link to the wiki at the top of this source module (will tack it on to the PR that fixes the marshalling issue).

Reply:
The wiki page covers what I was looking for, and with a link from the code I'm happy!

Reply:
The source may out-live a Github Wiki so moving the explanation to the implementation may be desirable.

Reply:
Yeah I'm not really sure what the best thing here is. It would be nice for the source files to be as self-contained as possible, and I definitely see the possibility of source and documentation diverging if they are kept separate, presenting a nonlocal relationship to be maintained.

On the other hand, there's value in such didactic documentation being accessible in "book" form without having to read source code, and there may be a balance to be struck.

Options:

- have the docs only on the wiki or only in the source
- duplicate docs in both places
- have implementation-specific docs (e.g. `generate-fused-operation`) in the source, and technical and theoretical material (e.g. what is stream fusion, here are some resources, here's how we use it at a high level in terms of e.g. producers transfomers and consumers, here where this is reflected in the source code)

The third option will likely entail some overlap between the docs, but as they reflect a different orientation it's possible the overlap would not be very significant (and they could mutually delegate to one another). If we think this is a good option, we could take the existing docs from the compiler wiki and distribute it along these lines.

Re: outlive, we could culturally commit to maintaining a wiki alongside the code, and augment our processes to support that in a sustainable way. E.g. we could add a bullet point in the PR template to say something like, "Have you updated the wiki (if applicable) to reflect the changes in your PR?"

We would still need to have links to the wiki if we went with this option, and I'm not sure how we'd avoid actual URLs, and actual references to GitHub in that case...

Related: #139

Reply:
The Git project maintains a glossary and other technical contribution material as part of the repository. They use plain-text, but scribble would do nicely and could be hosted as developer docs (perhaps as part of the qi-sdk package?).

Reply:
That sounds like a good option. From a wiki perspective, it would be ideal if (1) we had more styles for Scribble that we could customize to give it a wiki-like feel (e.g. "edit" links like those on the [Python Developer's Guide](https://devguide.python.org/developer-workflow/lang-changes/#langchanges)), (2) it recognized simplified wiki-style links resembling `[[a]]` and `[[a|b]]`, and (3) if we could edit and preview it via a web UI (e.g. Wikipedia/Github wiki, maybe Readthedocs). If we could have at least some of these that would make Scribble even more compelling vs Markdown/RST for this purpose. But I really don't know what the level of effort on these would be. Would it make sense to use Pollen?

Reply:
Section titles should have clickable edit links. Otherwise maybe you just want a collection of Markdown documents?

Reply:
Interesting, you mean this, when we click the heading?

```
Link to this section with
 @secref["Using_Qi_from_the_Host_Language"
         #:doc '(lib "qi/scribblings/qi.scrbl")]
Document source
 https://github.com/drym-org/qi/tree/main/qi-doc/scribblings/qi.scrbl
```

I forgot about this and it's quite functional. I think it's almost there, but it would be better if it was further simplified / specialized to a self-contained wiki, where there would be an explicit [Edit] button that simply follows the link to the Scribble source file, and where "link to this section" is unnecessary since we need not support cross-linking to other docs (so all links, except HTTP URLs, are internal, allowing the link syntax to be simple like [[a]] or [[a|b]]). I feel we especially want the developer docs to be [convenient and fast to edit](https://github.com/drym-org/qi/wiki/Writing-Developer-Docs).

We might be looking for a Scribble-based language here, comprising a set of macros/reader extensions + a CSS file. "Scribwiki"? 😆 But I'd love to learn more about Pollen to see if it's already usable in this way.

I'm also not sure if we'd want to move the entire wiki to Scribble or only the Developer's Guide.

Reply:
Also the link to the source file is just the root scribble file and not the specific one containing the content. Is that a bug?

Reply:
> Also the link to the source file is just the root scribble file and not the specific one containing the content. Is that a bug?

I think it's a regrettable consequence of `include-section`. Let me ask around.